mlir-tensorrt/tensorrt/lib/TensorRT/Transforms/TransposeReshapeElimination.cpp

```diff
@@ -0,0 +1,2568 @@
+          einsumEquationLhsParts[i] += (char)equation[j];
+        }
+        newInputs.push_back(transpose.getInput());
+      } else {
+        newInputs.push_back(input);
+      }
+    }
+
+    std::string newEinsumEquation = einsumEquationLhsParts[0];
+    for (int i = 1; i < einsumEquationLhsParts.size(); i++) {
+      newEinsumEquation += "," + einsumEquationLhsParts[i];
+    }
+    newEinsumEquation += "->" + einsumEquationRhs;
+
+    rewriter.replaceOpWithNewOp<tensorrt::EinsumOp>(op, op.getType(), newInputs,
+                                                    newEinsumEquation);
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+// Push up transpose from an einsum, rearranging the axes of the output tensor
+// in the einsum as needed transpose(einsum(x1, x2, ...)) -> einsum(x1, x2, ...)
+class PushUpTransposeToEinsum : public OpRewritePattern<tensorrt::TransposeOp> {
+public:
+  using OpRewritePattern<tensorrt::TransposeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensorrt::TransposeOp op,
+                                PatternRewriter &rewriter) const override {
+    auto perm = op.getPermutation();
+    if (!perm.isPermutation())
+      return failure();
+
+    auto einsum = op.getInput().getDefiningOp<tensorrt::EinsumOp>();
+    if (!einsum)
+      return failure();
+
+    SmallVector<OpOperand> einsumUsers(einsum->getUsers());
+    if (einsumUsers.size() != 1)
+      return failure();
+
+    std::string einsumEquation{einsum.getEquation()};
+    size_t pos = einsumEquation.find("->");
+    if (pos == std::string::npos)
+      return failure();
+    std::string einsumEquationLhs = einsumEquation.substr(0, pos);
+    std::string einsumEquationRhs = einsumEquation.substr(pos + 2);
+
+    SmallVector<int64_t> einsumRhs;
+    for (char c : einsumEquationRhs) {
+      einsumRhs.push_back(c);
+    }
+    einsumRhs = perm.compose(einsumRhs);
+    einsumEquationRhs = "";
+    for (int i = 0; i < einsumRhs.size(); i++) {
+      einsumEquationRhs += (char)einsumRhs[i];
+    }
+
+    std::string newEinsumEquation =
+        einsumEquationLhs + "->" + einsumEquationRhs;
+
+    auto newEinsum = rewriter.create<tensorrt::EinsumOp>(
+        op.getLoc(), op.getType(), einsum.getInputs(), newEinsumEquation);
+    rewriter.replaceOp(op, newEinsum.getResult());
+    rewriter.eraseOp(einsum);
+    return success();
+  }
+};
+} // namespace
+
+namespace {
+// Create an new transpose op from an einsum.  Rearrange the output axes to
+// match the ordering of the input axes This should enable converting the einsum
+// back to a matmul einsum(x1, x2, ...) -> transpose(einsum(x1, x2, ...))
+class EinsumPushDownTranspose : public OpRewritePattern<tensorrt::EinsumOp> {
+public:
+  using OpRewritePattern<tensorrt::EinsumOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(tensorrt::EinsumOp op,
+                                PatternRewriter &rewriter) const override {
+    for (auto input : op.getInputs()) {
```

Comment:
It's not clear what is the utility of this pattern, especially since one above absorbs the transpose.

It's not clear at all how you're reasoning about the cost of adding a transpose where it doesn't necessarily need to exist.

Reply:
NVM I see now below that this is related to raising einsum-to-matmul. I made another longer comment about questions with this approach.